goat.1
.\"
.\" SPDX-License-Identifier: BSD-2-Clause-FreeBSD
.\"
.\" Copyright (c) 2018 Mateusz Piotrowski <0mp@FreeBSD.org>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd June 12, 2018
.Dt GOAT 1
.Os
.Sh NAME
.Nm goat
.Nd POSIX-compliant shell movement boosting hack for real ninjas
.Sh SYNOPSIS
.Nm
.Op Ar command
.Sh DESCRIPTION
.Nm
is a utility to manage alias-like shortcuts to be used with
.Xr cd 1 .
.\" TODO: Add a note about libgoat.
.\" TODO: Add a note about cd.
.\" TODO: Add a note about shell rc files.
.Pp
The commands are as follows:
.Bl -tag -width "deleteprefix"
.It Cm delete Ar shortcut ...
Delete
.Ar shortcut
from
.Ev GOAT_PATH .
.It Cm deleteprefix Ar prefix
Delete all the shortcuts associated with directories, which start with the
given
.Ar prefix .
.It Cm fix
Remove shortcuts, which point to non-existing directories.
.It Cm help
Show help.
.It Cm link Ar shortcut Ar destination
Create a symbolic link between
.Ar shortcut
and
.Ar destination .
The link is going to be created in the directory pointed by the
.Ev GOAT_PATH
environmental variable.
The
.Cm link
keyword is usually optional, i.e.,
.Dq Li goat shortcut dest
is sufficient if it doesn't conflict with other
.Nm
commands.
The link is created for the absolute path of the
.Ar destination
directory.
Always succeeds.
.It Cm list
Show a list of all shortcuts.
.It Cm nuke
Delete all the shortcuts.
.It Cm version
Show version.
.El
.Sh GLOSSARY
.Bl -tag -width shortcut
.It Em destination
A directory pointed to by a
.Em shortcut .
.It Em link
A relation based on symbolic links between a
.Em shortcut
and a
.Em destination .
.It Em shortcut
A symbolic link file stored in
.Ev GOAT_PATH
that is associated with a certain
.Em destination .
When its name is specified as an argument to
.Xr cd 1
it causes
.Nm
to change the working directory to the
.Em destination
associated with the
.Em shortcut .
.El
.Sh ENVIRONMENT
.Bl -tag -width "GOAT_PATH" -compact
.It Ev GOAT_PATH
The directory where
.Nm
shortcuts are stored.
.El
.Sh EXAMPLES
Create a shortcut named
.Dq f
to
.Pa ~/Documents/dev/freebsd
(no need to use
.Cm link
here):
.Bd -literal -offset indent
goat f ~/Documents/dev/freebsd
.Ed
.Pp
Follow a link to change a directory with
.Xr cd 1 :
.Bd -literal -offset indent
cd f
.Ed
.Pp
Take the
.Dq f
shortcut and enter its destination subdirectory with just one command:
.Bd -literal -offset indent
cd f/ports
.Ed
.Pp
Create a shortcut named
.Dq p
to the current directory:
.Bd -literal -offset indent
goat p .
.Ed
.Pp
Go up the filesystem tree with
.Cm ...
.Po
same as the standard
.Dq Li cd ../../
.Pc :
.Bd -literal -offset indent
cd ...
.Ed
.Pp
List all your links:
.Bd -literal -offset indent
goat list
.Ed
.Pp
Delete a link (or more):
.Bd -literal -offset indent
goat delete h4xdir freebsd
.Ed
.Pp
Delete all the links which point to directories with the given prefix:
.Bd -literal -offset indent
goat deleteprefix $HOME/Documents
.Ed
.Sh COMPATIBILITY
.Nm
has no known compatibility issues.
.Pp
.Nm
should work just fine with MinGW environments.
.Sh SEE ALSO
.Xr cd 1 ,
.Xr ln 1 ,
.Xr sh 1
.Ss Similar software
.Bl -ohang
.It Xr autojump 1
.Lk https://github.com/wting/autojump
.It Xr fasd 1
.Lk https://github.com/clvv/fasd
.It Xr z 1
.Lk https://github.com/rupa/z/
.El
.Sh AUTHORS
.An Mateusz Piotrowski Aq Mt 0mp@FreeBSD.org
.Sh BUGS
Please report bugs at
.Lk https://github.com/0mp/goat/


src/goat.sh
#!/bin/sh
# vim: ts=2 sw=2 sts=2:
#
# goat - A hacky shortcut utility for managing your shortcuts.
#
# Notes:
# - Don't put tabs in the names of your shortcuts, files and directories as it
#   will break the shortcut file.
#

readonly kSHORTCUTS_DIR="${kINSTALLATION_DIR:-"$(dirname "$0")"}"
readonly kSHORTCUTS="$kSHORTCUTS_DIR/$kSHORTCUTS_FILE"

set -eu

gEXIT_STATUS=0
gPROPOSED_PATH=''
gRESOLVED_PATH=''
gSHORTCUT=''

# TODO
gt_error() {
  >&2 printf "ERROR: %s\n" "$1"
}

# TODO
gt_warn() {
  >&2 printf "WARNING: %s\n" "$1"
}

gt_info() {
  >&2 printf "INFO: %s\n" "$1"
}

gt_echo() {
  printf "%s" "$1"
}

# Remove an absolute shortcut from a file. Do nothing if there is no such an
# absolute shortcut
# Variables:
#   $1 -> a shortcut for which we want to remove a matching entry
# TODO
gt_remove() {
  if sed -e "/^$gSHORTCUT[[:space:]]\//d" "$kSHORTCUTS" > "$kSHORTCUTS.bak"; then
    mv "$kSHORTCUTS.bak" "$kSHORTCUTS"
    gt_info "goat removed the '$gSHORTCUT' shortcut"
    return 0
  else
    gt_info "goat didn't remove '$gSHORTCUT'; maybe it doesn't exist?"
    return 1
  fi
}

# Show usage information
# TODO
gt_usage() {
  cat <<EOF
usage: goat [<shortcut> | <shortcut> <path> | please [<arg> ...]]
  options:
    <shortcut> <directory>
      -- create a <shortcut> to a <directory>
    <shortcut>
      -- go to a directory assigned to <shortcut>; if a <shortcut> consists of
         dots only then it will expand every '.' to '../'
  pleases:
    list shortcuts
      -- list all your saved shortcuts
    edit shortcuts
      -- edit all your saved shortcuts in an editor
    nuke shortcuts
      -- delete all saved shortcuts
    delete <shortcut>
      -- delete <shortcut> from your saved shortcuts
    help me
      -- print this help message
EOF
  gEXIT_STATUS=1
  return 1
}

# TODO
gt_resolve_as_absolute() {
  candidatepaths="$(grep -n "^$(gt_echo "$gSHORTCUT" | sed "s/\./\\\./g")	/" \
    "$kSHORTCUTS")"
  if [ -z "$candidatepaths" ]; then
    return 1
  fi
  case "$(printf "%s\n" "$candidatepaths" | wc -l | tr -d '[:space:]')" in
    '1')
      gRESOLVED_PATH="$(gt_echo "$candidatepaths" | cut -f2)"
      return 0
      ;;
    '0')
      return 1
      ;;
    '')
      return 1
      ;;
    *)
      gt_warn "goat found more than one destinations for the '$gSHORTCUT'" \
        "shortcut; the '$kSHORTCUTS' file is corrupted"
      return 1
      ;;
  esac
}

# TODO
gt_resolve_as_relative() {
  candidatepaths="$(grep -n "^$gSHORTCUT	\." "$kSHORTCUTS")"
  case "$(gt_echo "$candidatepaths" | wc -l | tr -d '[:space:]')" in
    '1')
      gRESOLVED_PATH="$(gt_echo "$candidatepaths" | cut -d"\t" -f2)"
      return 0
      ;;
    '0')
      return 1
      ;;
    *)
      gt_echo "$candidatepaths" | while read -r candidate; do
        gRESOLVED_PATH="$(gt_echo "$candidate" | cut -d"\t" -f2)"
        if [ -d "$gRESOLVED_PATH" ]; then
          return 0
        fi
      done
      gRESOLVED_PATH=''
      return 1
      ;;
  esac
}

# TODO
gt_resolve_as_dots() {
  if [ "$gSHORTCUT" != "$(gt_echo "$gSHORTCUT" | tr -d -c '.')" ]; then
    return 1
  fi

  # Now we know there are only dots in the shortcut.
  gRESOLVED_PATH="$(gt_echo "$gSHORTCUT" | sed 's/\.//' | sed 's/\./\.\.\//g')"

  if [ -z "$gRESOLVED_PATH" ]; then
    return 1
  fi
}


# Go to the destination.
# The order of the destination look up is the following:
#  - absolute shortcuts (e.g. /usr/bin)
#  - relatvie shortcuts (e.g. ../../test) and stop when you find a working one
#  - dot shortcuts (e.g. ...)
# Input:
#   $1 -> A shortcut which can be matched with a directory we want to go to.
# Returns:
#   - a path when we found a coresponding path for the given shortcut
#   - 1 when there was an error
gt_go() {
  gSHORTCUT="$1"

  # Try absolute shortcuts.
  if gt_resolve_as_absolute; then
    :
  elif gt_resolve_as_relative; then
    :
  elif gt_resolve_as_dots; then
    :
  else
    gt_error "goat doesn't know where '$gSHORTCUT' should lead to"
    gEXIT_STATUS=1
    return 1
  fi

  if [ ! -d "$gRESOLVED_PATH" ]; then
    gt_error "goat cannot reach the '$gRESOLVED_PATH' directory"
    gEXIT_STATUS=1
    return 1
  fi

  printf "%s\n" "$gRESOLVED_PATH"
}

# Create a new shortcut
# Variables:
#   $1 -> a new shortcut
#   $2 -> a path for the new shortcut
# Returns:
#   - 1 if there was an error
#   - 0 if everything was fine
# TODO
gt_create() {
  gSHORTCUT="$1"
  gPROPOSED_PATH="$2"

  if [ ! -d "$gPROPOSED_PATH" ]; then
    gt_error "goat thinks that '$gPROPOSED_PATH' isn't a valid path"
    gEXIT_STATUS=1
    return 1
  fi

  # Get the absolute path
  gRESOLVED_PATH="$(cd -- "$gPROPOSED_PATH" && pwd -P 2>/dev/null | pwd)"

  # Remove the old shortcut
  gt_remove 2>/dev/null

  # Add new shortcut
  printf "%s	%s\n" "$gSHORTCUT" "$gRESOLVED_PATH" >> "$kSHORTCUTS"

  return 0
}

# Administrate your goat
# Variables:
#   $1 -> "please"
#   $2, $3 -> depends
# Returns:
#   0, if everything went fine
#   1, if there was an error
gt_configure() {
  if [ ! "$1" = "please" ]; then
    printf "FAUX PAS: goat doesn't like that kind of tone\n" >&2
    gEXIT_STATUS=1
    return 1
  fi

  case "$2" in
    'delete')
      gSHORTCUT="$3"
      gt_remove || gEXIT_STATUS=1
      ;;
    'edit')
      if [ "$3" = "shortcuts" ]; then
        # shellcheck disable=SC2094
        $kEDITOR "$kSHORTCUTS" < "$(tty)" > "$(tty)" || gEXIT_STATUS=1
      fi
      ;;
    'help')
      if [ "$3" = "me" ]; then
        gt_usage
      fi
      ;;
    'list')
      if [ "$3" = "shortcuts" ]; then
        cat "$kSHORTCUTS" || gEXIT_STATUS=1
      fi
      ;;
    'nuke')
      if [ "$3" = "shortcuts" ]; then
        rm "$kSHORTCUTS" || gEXIT_STATUS=1
      fi
      ;;
    *)
      gt_error "goat just doesn't know what to do with itself"
      gEXIT_STATUS=1
      ;;
  esac

  return $gEXIT_STATUS
}

# Handle user input.
# Creates the shortcuts file if it does not exists.
gt_run() {
  # Test existance of the file with shortcuts
  [ -d "$kSHORTCUTS" ] || mkdir -p "$kSHORTCUTS_DIR"
  [ -f "$kSHORTCUTS" ] || touch "$kSHORTCUTS"

  case "$#" in
    '0')
      gt_usage
      ;;
    '1')
      gt_go "$1"
      ;;
    '2')
      gt_create "$1" "$2"
      ;;
    '3')
      gt_configure "$1" "$2" "$3"
      ;;
  esac

  exit "$gEXIT_STATUS"
}

gt_run "$@"
